pkg/executor/sortexec/parallel_sort_spill_helper.go

```diff
@@ -0,0 +1,225 @@
+// Copyright 2024 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sortexec
+
+import (
+	"sync"
+	"sync/atomic"
+
+	"github.com/pingcap/tidb/pkg/types"
+	"github.com/pingcap/tidb/pkg/util"
+	"github.com/pingcap/tidb/pkg/util/chunk"
+	"github.com/pingcap/tidb/pkg/util/logutil"
+	"go.uber.org/zap"
+)
+
+type parallelSortSpillHelper struct {
+	cond             *sync.Cond
+	spillStatus      int
+	sortedRowsInDisk []*chunk.DataInDiskByChunks
+	sortExec         *SortExec
+
+	lessRowFunc   func(chunk.Row, chunk.Row) int
+	errOutputChan chan rowWithError
+	finishCh      chan struct{}
+
+	fieldTypes    []*types.FieldType
+	tmpSpillChunk *chunk.Chunk
+
+	bytesConsumed atomic.Int64
+	bytesLimit    atomic.Int64
+}
+
+func newParallelSortSpillHelper(sortExec *SortExec, fieldTypes []*types.FieldType, finishCh chan struct{}, lessRowFunc func(chunk.Row, chunk.Row) int, errOutputChan chan rowWithError) *parallelSortSpillHelper {
+	return &parallelSortSpillHelper{
+		cond:          sync.NewCond(new(sync.Mutex)),
+		spillStatus:   notSpilled,
+		sortExec:      sortExec,
+		lessRowFunc:   lessRowFunc,
+		errOutputChan: errOutputChan,
+		finishCh:      finishCh,
+		fieldTypes:    fieldTypes,
+		tmpSpillChunk: chunk.NewChunkWithCapacity(fieldTypes, spillChunkSize),
+	}
+}
+
+func (p *parallelSortSpillHelper) isNotSpilledNoLock() bool {
+	return p.spillStatus == notSpilled
+}
+
+func (p *parallelSortSpillHelper) isInSpillingNoLock() bool {
+	return p.spillStatus == inSpilling
+}
+
+func (p *parallelSortSpillHelper) isSpillNeeded() bool {
+	p.cond.L.Lock()
+	defer p.cond.L.Unlock()
+	return p.spillStatus == needSpill
```

Comment:
how about using atomic_int?

Reply:
> how about using atomic_int?

I think lock is better, as in `parallelSortSpillAction::Action` function, `sync.Cond` has been used to synchronize the spill and other goroutines so we already have lock and it's unnecessary to use atomic type.

Reply:
ok